src/releases/semver/version.rs

```diff
@@ -0,0 +1,140 @@
+use std::fmt::Display;
+use std::str::FromStr;
+
+use serde::{Deserialize, Serialize};
+
+use crate::step::StepError;
+
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub(crate) struct Version {
+    pub(crate) major: u64,
+    pub(crate) minor: u64,
+    pub(crate) patch: u64,
+    pub(crate) pre: Option<Prerelease>,
+}
+
+impl Ord for Version {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.major
+            .cmp(&other.major)
+            .then(self.minor.cmp(&other.minor))
+            .then(self.patch.cmp(&other.patch))
+            .then({
+                match (&self.pre, &other.pre) {
+                    (Some(pre), Some(other_pre)) => pre.cmp(other_pre),
+                    (Some(_), None) => std::cmp::Ordering::Less,
+                    (None, Some(_)) => std::cmp::Ordering::Greater,
+                    (None, None) => std::cmp::Ordering::Equal,
+                }
+            })
+    }
+}
+
+impl PartialOrd for Version {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl FromStr for Version {
+    type Err = StepError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let (version, pre) = s
+            .split_once('-')
+            .map_or((s, None), |(version, pre)| (version, Some(pre)));
+        let version_parts = version
+            .split('.')
+            .map(|part| {
+                part.parse::<u64>()
+                    .map_err(|err| StepError::InvalidSemanticVersion(err.to_string()))
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+        if version_parts.len() != 3 {
+            return Err(StepError::InvalidSemanticVersion(
+                "Version must have 3 parts".to_string(),
+            ));
+        }
+        Ok(Self {
+            major: version_parts[0],
+            minor: version_parts[1],
+            patch: version_parts[2],
+            pre: pre.map(Prerelease::from_str).transpose()?,
+        })
+    }
+}
+
+impl Display for Version {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}.{}.{}", self.major, self.minor, self.patch)?;
+        if let Some(pre) = &self.pre {
+            write!(f, "-{pre}")?;
+        }
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub(crate) struct Prerelease {
+    pub(crate) label: Label,
+    pub(crate) version: u64,
+}
+
+impl Display for Prerelease {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}.{}", self.label, self.version)
```

Comment:
semver.org describes the prerelease section as

> A pre-release version MAY be denoted by appending a hyphen and a series of dot separated identifiers immediately following the patch version... Examples: 1.0.0-alpha, 1.0.0-alpha.1, 1.0.0-0.3.7, 1.0.0-x.7.z.92, 1.0.0-x-y-z.--.

Not necessarily worth doing anything about for this PR, since `knope` defines the new versions, but potentially worth tracking. The current handling of `Prerelease` only supports a subset of what is allowed by semver.

Reply:
That is why I deleted the semver crate and wrote my own—too much dancing around all of the formats allowed by SemVer to make it more reasonable to deal with. I don't think I want Knope to ever get more flexible (unless someone comes up with a really good reason), but this behavior should be documented...